libindy-crypto/src/bls/mod.rs

```diff
@@ -305,7 +305,30 @@ impl Bls {
     /// Bls::sign(&message, &sign_key).unwrap();
     /// ```
     pub fn sign(message: &[u8], sign_key: &SignKey) -> Result<Signature, IndyCryptoError> {
-        let point = Bls::_hash(message)?.mul(&sign_key.group_order_element)?;
+        let point = Bls::_hash_sha2(message)?.mul(&sign_key.group_order_element)?;
+        Ok(Signature {
+            point,
+            bytes: point.to_bytes()?
+        })
+    }
+
+    /// Signs the message and returns signature to provide proof of possession.
+    ///
+    /// # Arguments
+    ///
+    /// * `message` - Message to sign
+    /// * `sign_key` - Sign key
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use indy_crypto::bls::*;
+    /// let message = vec![1, 2, 3, 4, 5];
+    /// let sign_key = SignKey::new(None).unwrap();
+    /// Bls::sign_pop(&message, &sign_key).unwrap();
+    /// ```
+    pub fn sign_pop(message: &[u8], sign_key: &SignKey) -> Result<Signature, IndyCryptoError> {
+        let point = Bls::_hash_sha3(message)?.mul(&sign_key.group_order_element)?;
```

Comment:
Since `sign_pop` is generating a signature, bbstract this operation `<hash of message>.mul(&sign_key.group_order_element` into a separate function and use that function in both `sign` and `sign_pop`. Its better if the hash function can be passed as an argument (as type `Digest` or type `enum`). Better, generating a `POP` should be a method in the `Verkey` implementation

Reply:
Not sure about pop is method of Verkey. As creating POP is operation similar to signing, so it closer to sign key rather Verkey.

Reply:
This is the interface i imagind
```
impl Verkey {
  pub fn new(sig_key... {
  }
  
  pub fn gen_pop(&self) -> Sig {
  }
}
```

Or
```
impl Verkey {
  pub fn new(sig_key... {
  }
  
  pub fn new_with_pop(sig_key...) -> (Self, Signature) {
  }
}
```

Reply:
@lovesh 
```
impl Verkey {
  pub fn new(sig_key) -> Self {
  }
  
  pub fn gen_pop(&self) -> Signature {
  }
}
```
It's impossible because `gen_pop` requires `SignKey` too

Reply:
My bad, i forgot to mention `SignKey`, lets pass `SignKey` as an arg.
```
pub fn gen_pop(&self, sig_key) -> Signature {
}
```

Reply:
I believe the most reliable way is to introduce PoP as a dedicated struct:

```Rust
struct ProofOfPossession
ProoOfPosestion::new(&VerKey, &SignKey)-> ProofOfPossession
BLS::verify_proof_of_posession(&VerKey, &PoP)
```

Reply:
Also i don't like struct BLS (it is a bad practice in rust as requires more complex imports). I suggest to make all BLS methods top level in bls module.